src/NHibernate.Test/NHSpecificTest/GH2750/FixtureByCode.cs

```diff
@@ -0,0 +1,173 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Transactions;
+using NHibernate.Cfg;
+using NHibernate.Cfg.MappingSchema;
+using NHibernate.Mapping.ByCode;
+using NUnit.Framework;
+
+namespace NHibernate.Test.NHSpecificTest.GH2750
+{
+	[TestFixture]
+	public class ByCodeFixture : TestCaseMappingByCode
+	{
+		protected override HbmMapping GetMappings()
+		{
+			var mapper = new ModelMapper();
+			mapper.Class<TestEntity>(rc => { rc.Id(x => x.Id, m => m.Generator(Generators.Assigned)); });
+
+			return mapper.CompileMappingForAllExplicitlyAddedEntities();
+		}
+
+		protected override void Configure(Configuration configuration)
+		{
+			configuration.SetProperty(Cfg.Environment.BatchSize, 10.ToString());
+			configuration.SetProperty(Cfg.Environment.UseConnectionOnSystemTransactionPrepare, true.ToString());
+		}
+
+		protected override void OnTearDown()
+		{
+			using (var session = OpenSession())
+			using (var transaction = session.BeginTransaction())
+			{
+				session.CreateQuery("delete from System.Object").ExecuteUpdate();
+
+				transaction.Commit();
+			}
+		}
+
+		[Test]
+		public void ShouldWorkWithOuterSystemTransaction()
+		{
+			const int count = 3;
+			var entities = new List<TestEntity>(count);
+			for (var i = 0; i < count; i++)
+			{
+				entities.Add(new TestEntity { Id = Guid.NewGuid() });
+			}
+
+			using (var transaction = new TransactionScope())
+			{
+				using (var session = Sfi.OpenStatelessSession())
+				{
+					foreach (var entity in entities)
+					{
+						session.Insert(entity);
+					}
+				}
+
+				transaction.Complete();
+			}
+
+			using (var session = OpenSession())
+			{
+				var results = session.QueryOver<TestEntity>().List<TestEntity>();
+
+				Assert.That(results.Count, Is.EqualTo(count));
+			}
+		}
+
+		[Test]
+		public void ShouldWorkWithInnerSystemTransaction()
+		{
+			const int count = 3;
+			var entities = new List<TestEntity>(count);
+			for (var i = 0; i < count; i++)
+			{
+				entities.Add(new TestEntity { Id = Guid.NewGuid() });
+			}
+
+			using (var session = Sfi.OpenStatelessSession())
+			{
+				using (var transaction = new TransactionScope())
+				{
+					foreach (var entity in entities)
+					{
+						session.Insert(entity);
+					}
+					session.FlushBatcher();
```

Comment:
Is there any reason why we cannot just call `_session.Flush()` in `SystemTransactionContext.Prepare(PreparingEnlistment)`?

Reply:
Yes we cannot. That is too late, we are already in the 2PC, exposed to threading concurrency issues if trying to do anything that will use the session current connection.

At this point, the "use connection on system transaction" setting is already required for allowing attempt to use a connection in the 2PC. So anyway if this setting is not enabled (as recommended in the [documentation](https://nhibernate.info/doc/nhibernate-reference/transactions.html#transactions-scopes)), we cannot flush here.
But in the case of stateless session we cannot flush even if the setting is enabled, because if there is anything to flush in the batcher, it has already acquired the session current connection. And so the connection switch done in `ConnectionManager.BeginProcessingFromSystemTransaction` will have no effect on the batcher, causing it to flush on the session connection which may be executing its own 2PC on another thread at the same time.

There no way to be safe here excepted the [recommend documented one](https://nhibernate.info/doc/nhibernate-reference/transactions.html#transactions-scopes):
> For new applications, it is recommended to set transaction.use_connection_on_system_prepare to false, and to flush explicitly your sessions before scope completion.